modules/core/lib/Controller/Login.php

```diff
@@ -43,11 +64,391 @@ public function __construct(
         $this->config = $config;
     }
 
-    public function welcome(): Response
+
+    /**
+     * Inject the \SimpleSAML\Auth\Source dependency.
+     *
+     * @param \SimpleSAML\Auth\Source $authSource
+     */
+    public function setAuthSource(Auth\Source $authSource): void
+    {
+        $this->authSource = $authSource;
+    }
+
+
+    /**
+     * Inject the \SimpleSAML\Auth\State dependency.
+     *
+     * @param \SimpleSAML\Auth\State $authState
+     */
+    public function setAuthState(Auth\State $authState): void
+    {
+        $this->authState = $authState;
+    }
+
+
+    /**
+     * @return \SimpleSAML\XHTML\Template
+     */
+    public function welcome(): Template
     {
         return new Template($this->config, 'core:welcome.twig');
     }
 
+
+    /**
+     * This page shows a username/password login form, and passes information from it
+     * to the \SimpleSAML\Module\core\Auth\UserPassBase class, which is a generic class for
+     * username/password authentication.
+     *
+     * @param \Symfony\Component\HttpFoundation\Request $request
+     * @return \SimpleSAML\XHTML\Template
+     */
+    public function loginuserpass(Request $request): Template
+    {
+        // Retrieve the authentication state
+        if (!$request->query->has('AuthState')) {
+            throw new Error\BadRequest('Missing AuthState parameter.');
+        }
+        $authStateId = $request->query->get('AuthState');
```

Comment:
Should we do some kind of basic validation on this input variable before we try to load a state with it?

Reply:
I don't mind if we do, but we never did any of that before..